lib/orderbook/OrderBook.ts

```diff
@@ -299,42 +321,76 @@ class OrderBook extends EventEmitter {
       const remainingOrder: StampedOwnOrder = result.remainingOrder || { ...order, quantity: 0 };
       swapFailures.forEach(order => remainingOrder.quantity += order.quantity);
 
-      // invoke addOwnOrder recursively, append matches/swaps and set the consecutive remaining order
-      const remainingOrderResult = await this.addOwnOrder(remainingOrder, false, onUpdate, maxTime);
+      // invoke placeOrder recursively, append matches/swaps and set the consecutive remaining order
+      const remainingOrderResult = await this.placeOrder(remainingOrder, false, onUpdate, maxTime);
       result.internalMatches.push(...remainingOrderResult.internalMatches);
       result.swapResults.push(...remainingOrderResult.swapResults);
       result.remainingOrder = remainingOrderResult.remainingOrder;
     }
 
     const { remainingOrder } = result;
     if (remainingOrder && !discardRemaining) {
-      matchingEngine.addOwnOrder(remainingOrder);
-      this.localIdMap.set(remainingOrder.localId, { id: remainingOrder.id, pairId: remainingOrder.pairId });
-      this.emit('ownOrder.added', remainingOrder);
-      this.logger.debug(`order added: ${JSON.stringify(remainingOrder)}`);
-
-      this.broadcastOrder(remainingOrder);
+      this.addOwnOrder(remainingOrder);
       onUpdate && onUpdate({ case: PlaceOrderEventCase.RemainingOrder, payload: remainingOrder });
     }
 
     return result;
   }
 
+  public executeSwap = async (orderId: string, pairId: string, peerPubKey: string, quantity?: number): Promise<SwapResult> => {
+    const maker = this.removePeerOrder(orderId, pairId, peerPubKey, quantity).order;
+    const taker = this.stampOwnOrder({
+      pairId,
+      localId: '',
+      price: maker.price,
+      isBuy: !maker.isBuy,
+      quantity: quantity || maker.quantity,
+    });
```

Comment:
What I'm thinking here is I'd put the logic that creates `maker` and `taker` back in Service.ts. These are public methods already, so I don't see a problem with that. Then the code below (the try/catch block) can stay in this method, with the method only taking `maker` and `taker` parameters.

Then replace the try/catch starting at line 306 with a call to this method, and in the catch block there you can have the `swapFailures.push` line.

Let me know what you think.

Reply:
That's what I was thinking of doing, but I thought that it would be confusing for `placeOrder` flow that some of the order removal events would be triggered from within `executeSwap`, so I preferred code duplication over that. We can still make `placeOrder` to emit these events, but then `executeSwap` will need another wrapper. 
In addition, `placeOrder` still haven't stabilized, so I didn't want to over-optimize. 

Tell me what you think, and if that's the only thing left, let's merge it and i'll open a separate PR for that.

Reply:
@moshababo To me it's not confusing, it's just that this method would be responsible for calling `swaps.executeSwap` and for emitting the `peerOrder.filled` and `peerOrder.invalidation` events. Then the `placeOrder` method will still be responsible for `swapResults.push`, `swapFailures.push` and the `onUpdate` callback (none of which are relevant when calling `executeSwap` via the RPC).

But yeah it's not a big deal so no need to get stuck on this - if you're still not sure leave as-is and I'll make a separate PR to discuss this specific issue. And I think the only other (minor) thing is the `Promise.resolve` comment, it's the only other unresolved conversation left but it's hidden so you have to expand it in the PR.

Reply:
Yes, I just thought that having this method emitting the events will be confusing when reading `placeOrder`, which also emits events.
I’ll revise this idea tomorrow, and perhaps open a separate PR. 

Saw your other comment and replied.